tests/test_adjoint.py

```diff
@@ -22,46 +22,46 @@
 
 import pytest
 import torch
+from . import utils
 
 import torchsde
+from torchsde.settings import NOISE_TYPES, METHODS
 from .basic_sde import BasicSDE1, BasicSDE2, BasicSDE3, BasicSDE4
-from .problems import Ex1, Ex2, Ex3
-from .utils import assert_allclose
+from .problems import Ex1, Ex2, Ex3, Ex4
 
 torch.manual_seed(1147481649)
 torch.set_default_dtype(torch.float64)
 device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
 dtype = torch.get_default_dtype()
 
-ito_methods = {'euler': 'ito',
-               'milstein': 'ito',
-               'srk': 'ito'}
-stratonovich_methods = {'midpoint': 'stratonovich',
-                        'log_ode': 'stratonovich'}
+ito_methods = {'milstein': 'ito', 'srk': 'ito'}
+stratonovich_methods = {'midpoint': 'stratonovich', 'log_ode': 'stratonovich'}
 
 
-@pytest.mark.parametrize("problem", [Ex1, Ex2, Ex3])
+@pytest.mark.parametrize("sde_cls", [Ex1, Ex2, Ex3, Ex4])
 @pytest.mark.parametrize("method, sde_type", itertools.chain(ito_methods.items(), stratonovich_methods.items()))
-@pytest.mark.parametrize("noise_type", ['diagonal', 'scalar', 'additive', 'general'])
-@pytest.mark.parametrize('adaptive', (False, True))
-def test_adjoint(problem, method, sde_type, noise_type, adaptive):
-    if method == 'euler' and adaptive:
-        return
-    if problem is not Ex3 and noise_type == 'additive':
+@pytest.mark.parametrize('adaptive', (False,))
+def test_adjoint(sde_cls, method, sde_type, adaptive):
+    # Skipping below, since method not supported for corresponding noise types.
+    if method == METHODS.log_ode_midpoint and sde_cls.noise_type == NOISE_TYPES.diagonal:
```

Comment:
So I've just checked, and indeed log-ODE states that it doesn't supports diagonal noise.
I think that's a mistake - once fixed this check can also be removed.

Reply:
SGTM.